core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
+            Self::kura_recieve_blocks_thread(&kura_weak_reference);
+        });
+        Ok(kura)
+    }
 
     /// Loads kura from configuration
     /// # Errors
     /// Fails if call to new fails
-    async fn from_configuration(
+    pub fn from_configuration(
         configuration: &config::KuraConfiguration,
         wsv: Arc<WorldStateView>,
         broker: Broker,
-    ) -> Result<Self> {
+    ) -> Result<Arc<Self>> {
         Self::new(
             configuration.init_mode,
             Path::new(&configuration.block_store_path),
-            configuration.blocks_per_storage_file,
             wsv,
             broker,
             configuration.actor_channel_capacity,
         )
-        .await
     }
-}
 
-#[async_trait]
-impl KuraTrait for Kura {
-    async fn new(
-        mode: Mode,
-        block_store_path: &Path,
-        blocks_per_file: NonZeroU64,
-        wsv: Arc<WorldStateView>,
-        broker: Broker,
-        mailbox: u32,
-    ) -> Result<Self> {
-        Self::new_meta(
-            mode,
-            block_store_path,
-            blocks_per_file,
-            wsv,
-            broker,
-            mailbox,
-            DefaultIO,
-        )
-        .await
+    /// After constructing [`Kura`] it should be initialized to be ready to work with it.
+    ///
+    /// # Errors
+    /// * May fail if file storage is either unavailable or data is invalid/corrupted
+    ///
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
+    pub fn init(&self) -> Result<Vec<VersionedCommittedBlock>> {
+        let mut blocks = Vec::new();
+
+        let block_store = self.block_store.lock().expect("lock block store");
+
+        let block_index_count: usize = block_store
+            .read_index_count()?
+            .try_into()
+            .expect("We don't have 4 billion blocks.");
+        let mut block_indices = vec![(0_u64, 0_u64); block_index_count];
+        block_store.read_block_indices(0, &mut block_indices)?;
+
+        for (block_start, block_len) in block_indices {
+            let mut block_data_buffer = vec![
+                0_u8;
+                block_len.try_into().expect(
+                    "TODO: handle allocation too large because block_len is corrupted."
+                )
+            ];
+            match block_store.read_block_data(block_start, &mut block_data_buffer) {
+                Ok(_) => match VersionedCommittedBlock::decode_versioned(&block_data_buffer) {
+                    Ok(decoded_block) => {
+                        blocks.push(decoded_block);
+                    }
+                    Err(error) => {
+                        error!(%error, "Encountered malformed block. Not reading any blocks beyond this height.");
+                        break;
+                    }
+                },
+                Err(error) => {
+                    error!(%error, "Malformed block index or corrupted block data file. Not reading any blocks beyond this height.");
```

Comment:
```suggestion
                    error!(?error, "Malformed block index or corrupted block data file. Not reading any blocks beyond this height.");
```

The debug representation contains the traceback. This will be useful once we have some more details to add to the error.